### PR TITLE
[SPARK-4033][Examples]Input of the SparkPi  too big causes the emption exception

### PR DESCRIPTION
**Pull request by SaintBacchus**
If input of the SparkPi args is larger than the 25000, the integer 'n' inside the code will be overflow, and may be a negative number.
And it causes  the (0 until n) Seq as an empty seq, then doing the action 'reduce'  will throw the UnsupportedOperationException("empty collection").

The max size of the input of sc.parallelize is Int.MaxValue - 1, not the Int.MaxValue.

### CODE REVIEW

**Comment by andrewor14** [comment]
@SaintBacchus Any update on this PR? As @JoshRosen I think the correct fix here is to use `Long` instead of `Int`. The current solution here probably works but adds unnecessary complexity to the supposably very easy to understand example. This is intended to be many users' first Spark application so I think keeping it simple is a very important consideration.

**Comment by SaintBacchus** [comment]
@andrewor14 I had explained why it can not use `Long` instead of `Int`. Not only the `Range` but also the `Partition` only can be appropriate with `Int`, and can't converse to a `Long`.
Can we restrict the input and log an error to exit from the process?

**Comment by andrewor14** [comment]
I see. I think it's simplest to just prompt the user to provide a smaller input if the input would cause an overflow. For instance: https://github.com/andrewor14/spark/compare/spark-pi-validate-args. I don't feel strongly for or against the alternative that @srowen suggested, but I personally find the fail-fast solution easiest to understand.

**Comment by andrewor14** [comment]
Actually, on second thought I think the existing solution in this PR is probably sufficient. Although this currently still fails if `slices <= 0`, that's kind of the user's problem and it will fail quickly. I'm OK with not printing a warning because it adds extra logic to this program that's supposed to be very small.

File: examples/src/main/scala/org/apache/spark/examples/SparkPi.scala [hunk 1]
```diff
@@ -27,7 +27,7 @@ object SparkPi {
     val conf = new SparkConf().setAppName("Spark Pi")
     val spark = new SparkContext(conf)
     val slices = if (args.length > 0) args(0).toInt else 2
-    val n = 100000 * slices
+    val n = math.min(100000L * slices, Int.MaxValue -1).toInt
```
Comment (andrewor14) [inline examples/src/main/scala/org/apache/spark/examples/SparkPi.scala]: yes, this can just be `Int.MaxValue`

Comment (andrewor14) [inline examples/src/main/scala/org/apache/spark/examples/SparkPi.scala]: maybe also add a quick comment at the end to explain why we're doing this:
```
val n = math.min(100000L * slices, Int.MaxValue).toInt // avoid overflow
```

**Comment by SaintBacchus** [comment]
Sorry for forgetting explain why there must be  `Int.MaxValue -1` instead of  `Int.MaxValue` .
As the SparkPi using the `parallelize` to generator the RDD so it has a limit of input seq, and go through the implement of this:
```scala
    def positions(length: Long, numSlices: Int): Iterator[(Int, Int)] = {
      (0 until numSlices).iterator.map(i => {
        val start = ((i * length) / numSlices).toInt
        val end = (((i + 1) * length) / numSlices).toInt
        (start, end)
      })
    }
``` 
in line 122  at `ParallelCollectionRDD.scala`
The max size of the input seq was `Int.MaxValue - 1`, otherwise the `end = (((i + 1) * length) / numSlices).toInt` will overflow and lead an uncorrect reslut.
I tested this:
```scala
  sc.makeRDD(1 to (Int.MaxValue)).count       // result = 0
  sc.makeRDD(1 to (Int.MaxValue - 1)).count   // result = 2147483646 = Int.MaxValue - 1
  sc.makeRDD(1 until (Int.MaxValue)).count    // result = 2147483646 = Int.MaxValue - 1
```

**Comment by advancedxy** [comment]
Hi @SaintBacchus, I get the idea that there is a limit in the seq size. But I don't think the buggy code is from implementation you listed. The actually buggy code is the first case in the pattern matching.

``` Scala
case r: Range.Inclusive => {
        val sign = if (r.step < 0) {
          -1
        } else {
          1
        }
        slice(new Range(
          r.start, r.end + sign, r.step).asInstanceOf[Seq[T]], numSlices) // r.end + sign is overflow
      }
```
We should consider fix that bug. @srowen and @andrewor14, do you think we need another pr?
@SaintBacchus do you think you have time for that? I can contribute my work if you are busy.

**Comment by srowen** [comment]
@advancedxy I do think it's a separate issue. The case of integer overflow could be handled better in both places.

**Comment by andrewor14** [comment]
I see, @advancedxy would you mind filing a JIRA for this? If it's convenient for you to do so, feel free to also create a PR for it.

As for this PR, the latest changes LGTM. The difference between `to` and `until` is trivial enough in this particular case that I think it's OK for this to go in. The existing alternative would have failed the application anyway, so I believe this piece of code is strictly better than before. Merging into master.

**Comment by advancedxy** [comment]
@andrewor14, of course, I will file a jira and send a pr when I get some spare time later today.

**Comment by SaintBacchus** [comment]
@advancedxy  I had showed the wrong bad code:sparkles: Thanks for  point it out

**Pull request closed without merging**